### PR TITLE
fix: add data_type and nullable to StructField hash (#2045)

### LINKED ISSUE #2045: Hash for StructField should consider more than the name
Two StructFields with the same name but different data_types are considered to hash equivalently, which I think is incorrect.

I'm working around this for now, but filing for @roeap and I to discuss later

### PR DESCRIPTION
**Pull request by sonhmai**
# Description
Correct hashing of StructField and add tests.
- Current:  Two StructFields with the same name but different `data_type` or different `nullable` are considered to hash equivalently.
- To be: `name`, `data_type` and `nullability` should match for a field to be considered equivalent.

# Related Issue(s)
closes #2045

### CODE REVIEW

**Comment by sonhmai** [comment]
there seems to be an error unrelated to this PR

```
FAILED tests/test_convert_to_delta.py::test_convert_delta_with_partitioning - _internal.DeltaError: 
Generic error: The schema of partition columns must be provided to convert a Parquet table to a Delta table
```

**Comment by ion-elgreco** [comment]
@sonhmai not so sure about that, these failures weren't there before. Maybe some interaction somewhere else in the code?

**Comment by sonhmai** [comment]
> @sonhmai not so sure about that, these failures weren't there before. Maybe some interaction somewhere else in the code?

yep. will definitely look into details.

**Comment by sonhmai** [comment]
@ion-elgreco  fix pushed with local tests passed. can you plz let workflow run? Thanks.

**Comment by sonhmai** [comment]
@rtyler  @ion-elgreco  all checks passed!

**Review by ion-elgreco — APPROVED** [review]
@sonhmai thanks for the fix! 😀

File: crates/core/src/kernel/models/schema.rs [hunk 1]
```diff
@@ -641,6 +643,7 @@ mod tests {
     use super::*;
     use serde_json;
     use serde_json::json;
+    use std::hash::DefaultHasher;
```
Comment (emcake) [inline crates/core/src/kernel/models/schema.rs:646]: incase anyone else runs into this - this raises the MSRV of running the tests to 1.76.

Comment (sonhmai) [reply crates/core/src/kernel/models/schema.rs:646]: thanks for pointing this out!

**Pull request merged**